### PR TITLE
[feat]: Add vae encoder embedded generator to main

### PR DESCRIPTION
**Pull request by rlsu9**
run `scripts/data_preprocess/finetune_data_gen.sh` to generate the pre processed data

### CODE REVIEW

**Comment by foreverpiano** [comment]
图片
please delete unsued file and use import diffuser

File: fastvideo/dataset/__init__.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-
+from torchvision.transforms import Lambda
 from fastvideo.dataset.t2v_datasets import T2V_dataset
 from fastvideo.dataset.latent_datasets import LatentDataset
 from fastvideo.dataset.transform import Normalize255, TemporalRandomCrop,CenterCropResizeVideo
 
 def getdataset(args):
     temporal_sample = TemporalRandomCrop(args.num_frames)  # 16 x
-    norm_fun = ae_norm[args.ae]
+    norm_fun = Lambda(lambda x: 2. * x - 1.)
```
Comment (jzhang38) [inline fastvideo/dataset/__init__.py:11]: Can you give a reference here?

File: fastvideo/dataset/t2v_datasets.py [hunk 2]
```diff
@@ -168,23 +169,25 @@ def get_data(self, idx):
...
+            assert False
             return self.get_image(idx)
     
     def get_video(self, idx):
 
         video_path = dataset_prog.cap_list[idx]['path']
         assert os.path.exists(video_path), f"file {video_path} do not exist!"
-        frame_indice = dataset_prog.cap_list[idx]['sample_frame_index']
-        video = self.decord_read(video_path, predefine_num_frames=len(frame_indice))
+        video, _, metadata = torchvision.io.read_video(video_path, output_format='THWC')
```
Comment (jzhang38) [inline fastvideo/dataset/t2v_datasets.py]: Why switch from decord to torchvision.io.read_video?

File: fastvideo/dataset/t2v_datasets.py [hunk 3]
```diff
@@ -168,23 +169,25 @@ def get_data(self, idx):
...
         h, w = video.shape[-2:]
         assert h / w <= 17 / 16 and h / w >= 8 / 16, f'Only videos with a ratio (h/w) less than 17/16 and more than 8/16 are supported. But video ({video_path}) found ratio is {round(h / w, 2)} with the shape of {video.shape}'
-        t = video.shape[0]
-        video = self.transform(video)  # T C H W -> T C H W
-
-        # video = torch.rand(221, 3, 480, 640)
-
-        video = video.transpose(0, 1)  # T C H W -> C T H W
+        
+        video = video.float() / 127.5 - 1.0
```
Comment (jzhang38) [inline fastvideo/dataset/t2v_datasets.py:200]: Where does this come from?

File: fastvideo/utils/dit/joint_model/layers.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            return tuple(x)
+        return tuple(repeat(x, n))
+
+    return parse
+
+
+to_2tuple = _ntuple(2)
+
+
+class TimestepEmbedder(nn.Module):
```
Comment (jzhang38) [inline fastvideo/utils/dit/joint_model/layers.py]: Please delete unnecessary files.

File: fastvideo/utils/dit/joint_model/context_parallel.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    if _CONTEXT_PARALLEL_GROUP is None:
+        raise RuntimeError("CP group not initialized")
+    return _CONTEXT_PARALLEL_GROUP
+
+
+def is_cp_active():
+    return _CONTEXT_PARALLEL_GROUP is not None
+
+
+class AllGatherIntoTensorFunction(torch.autograd.Function):
```
Comment (jzhang38) [inline fastvideo/utils/dit/joint_model/context_parallel.py]: Please delete unnecessary files.

File: fastvideo/dataset/t2v_datasets.py [hunk 6]
```diff
@@ -168,23 +169,25 @@ def get_data(self, idx):
...
+        video, _, metadata = torchvision.io.read_video(video_path, output_format='THWC')
+        fps = metadata["video_fps"]
+        video = rearrange(video, 't h w c -> c t h w')
+        video = video.unsqueeze(0)
+        assert video.dtype == torch.uint8
 
         h, w = video.shape[-2:]
         assert h / w <= 17 / 16 and h / w >= 8 / 16, f'Only videos with a ratio (h/w) less than 17/16 and more than 8/16 are supported. But video ({video_path}) found ratio is {round(h / w, 2)} with the shape of {video.shape}'
-        t = video.shape[0]
-        video = self.transform(video)  # T C H W -> T C H W
```
Comment (jzhang38) [inline fastvideo/dataset/t2v_datasets.py:183]: Why delete self.transform?

**Comment by jzhang38** [comment]
Can you add a script to decode back the latents to videos? So we can at least verify the correctness of the encoding pipeline.

**Comment by ghost** [comment]
"A person is sawing a piece of wood placed on a support, using a large hand saw. The saw moves rapidly back and forth across the top of the wood, generating small wood chips and dust. The person's hands are visible, gripping the saw firmly. The background shows a workshop environment with indistinct objects and a bench. The focus remains on the action of sawing, with no significant change in the camera angle or background. The sawing action continues with the saw moving through the wood, which shows a visible cut deepening as the saw progresses. Wood chips continue to scatter from the cutting area. The person's grip adjusts slightly for better control and force. The background remains consistent with the workshop setting, and the camera stays focused on the sawing action without any shifts in perspective. No new objects or persons enter the scene. The sawing action reaches a deeper cut into the wood, indicating progress in the task. The person's movements become more pronounced as they apply more force. The saw blade appears more prominently as it moves back and forth. The background and setting remain unchanged, maintaining the focus on the sawing activity. The camera continues to capture the action closely, emphasizing the interaction between the saw and the wood."

File: fastvideo/dataset/t2v_datasets.py [hunk 7]
```diff
@@ -168,23 +170,37 @@ def get_data(self, idx):
...
+        video, _, metadata = torchvision.io.read_video(video_path, output_format='TCHW')
+        video = self.transform(video) 
+        video = rearrange(video, 't c h w -> c t h w')
+        video = video.unsqueeze(0)
+        video =  video.to(torch.uint8)
+        assert video.dtype == torch.uint8
+        target_length = self.target_length
+        current_length = video.shape[2]  # This is t (92 in your example)
+
+        if current_length < target_length:
```
Comment (jzhang38) [inline fastvideo/dataset/t2v_datasets.py:188]: Can we remove those videos?

File: fastvideo/dataset/t2v_datasets.py [hunk 8]
```diff
@@ -168,23 +170,37 @@ def get_data(self, idx):
...
         else:
+            assert False
             return self.get_image(idx)
     
     def get_video(self, idx):
 
         video_path = dataset_prog.cap_list[idx]['path']
         assert os.path.exists(video_path), f"file {video_path} do not exist!"
-        frame_indice = dataset_prog.cap_list[idx]['sample_frame_index']
-        video = self.decord_read(video_path, predefine_num_frames=len(frame_indice))
```
Comment (jzhang38) [inline fastvideo/dataset/t2v_datasets.py:178]: Can you remove def decord_read ?

File: fastvideo/dataset/t2v_datasets.py [hunk 9]
```diff
@@ -277,10 +293,10 @@ def define_frame_index(self, cap_list):
                     hw_aspect_thr = 1.5
                     is_pick = filter_resolution(height, width, max_h_div_w_ratio=hw_aspect_thr*aspect, 
                                                 min_h_div_w_ratio=1/hw_aspect_thr*aspect)
-                    if not is_pick:
-                        print("resolution mismatch")
-                        cnt_resolution_mismatch += 1
-                        continue
+                    # if not is_pick:
```
Comment (jzhang38) [inline fastvideo/dataset/t2v_datasets.py:294]: Why do you remove this?

File: fastvideo/sample/vae_encoder_pipe.py [hunk 10]
```diff
@@ -0,0 +1,464 @@
+# Copyright 2024 Black Forest Labs and The HuggingFace Team. All rights reserved.
```
Comment (jzhang38) [inline fastvideo/sample/vae_encoder_pipe.py]: Mark

File: fastvideo/utils/data_preprocess/finetune_data_T5.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    video_in_tensor_json_path = os.path.join(args.output_dir, "videos2caption_temp.json")
+    with open(video_in_tensor_json_path, "r") as f:
+        train_dataset = json.load(f)
+    
+        json_data = []
+        for idx, data in enumerate(train_dataset):
+            # try:
+            with torch.inference_mode():
+                with torch.autocast("cuda", dtype=torch.bfloat16):
+                    video_in_tensor = torch.load(os.path.join(args.output_dir, 'video_in_tensor', data['video_in_tensor']))
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_T5.py]: video_in_tensor name

This is too hack.

File: fastvideo/utils/data_preprocess/finetune_data_T5.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+from accelerate.utils import DistributedType, ProjectConfiguration
+from fastvideo.utils.vae.models import Encoder
+from fastvideo.sample.vae_encoder_pipe import MochiEncoderPipeline
+from safetensors.torch import load_file
+from diffusers.utils import export_to_video
+import json
+import os
+logger = get_logger(__name__)
+
+def main(args):
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_T5.py]: Multi-process version?

File: fastvideo/utils/data_preprocess/finetune_data_VAE.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
+from fastvideo.dataset import getdataset
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_VAE.py:1]: Multi-process version.

File: fastvideo/utils/data_preprocess/finetune_data_VAE.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    json_data = []
+    for idx, data in enumerate(train_dataset):
+        # try:
+        with torch.inference_mode():
+            with torch.autocast("cuda", dtype=torch.bfloat16):
+                ldist = encoder(data['pixel_values'].to(encoder_device))
+
+                video_name = str(idx)
+                video_in_tensor = ldist.sample()
+                video_in_tensor_path = os.path.join(args.output_dir, "video_in_tensor", video_name + ".pt")
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_VAE.py]: video_in_tensor?

File: fastvideo/utils/data_preprocess/finetune_data_VAE.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
+from fastvideo.dataset import getdataset
+from torch.utils.data import DataLoader
+from fastvideo.utils.dataset_utils import Collate
+import argparse
+import torch
+from accelerate import Accelerator
+from accelerate.logging import get_logger
+from accelerate.utils import DistributedType, ProjectConfiguration
+from fastvideo.utils.vae.models import Encoder
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_VAE.py]: Where is it?

**Comment by jzhang38** [comment]
Please add README or .sh files on how to do the preprocessing.

File: fastvideo/dataset/__init__.py (hunk 1 above)
Comment (rlsu9) [reply fastvideo/dataset/__init__.py:11]: used to normalized the gaussian distribution

File: fastvideo/utils/data_preprocess/finetune_data_VAE.py (hunk 15 above)
Comment (rlsu9) [reply fastvideo/utils/data_preprocess/finetune_data_VAE.py]: fixed by changing gitignore

File: fastvideo/utils/data_preprocess/finetune_data_VAE.py (hunk 14 above)
Comment (rlsu9) [reply fastvideo/utils/data_preprocess/finetune_data_VAE.py]: fixed in commit # @[https://github.com/jzhang38/FastVideo-OSP/pull/30/commits/0f347227bb24ecb2a1a3385534b3c244390b57cc](url)

File: fastvideo/utils/data_preprocess/finetune_data_VAE.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    local_rank = int(os.getenv('RANK', 0))
+    world_size = int(os.getenv('WORLD_SIZE', 1))
+    print('world_size', world_size, 'local rank', local_rank)
+    args.ae_stride_t, args.ae_stride_h, args.ae_stride_w = 4, 8, 8
+    args.ae_stride = args.ae_stride_h
+    patch_size_t, patch_size_h, patch_size_w = 1, 2, 2
+    args.patch_size = patch_size_h
+    args.patch_size_t, args.patch_size_h, args.patch_size_w = patch_size_t, patch_size_h, patch_size_w
+    accelerator_project_config = ProjectConfiguration(project_dir=args.output_dir, logging_dir=args.logging_dir)
+    accelerator = Accelerator(
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_VAE.py:26]: Why do you need accelerator?

File: fastvideo/utils/data_preprocess/finetune_data_T5.py [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    pipe = MochiPipeline.from_pretrained(args.model_path, torch_dtype=torch.bfloat16).to(device)
+    pipe.vae.enable_tiling()
+    os.makedirs(args.output_dir, exist_ok=True)
+    os.makedirs(os.path.join(args.output_dir, "video"), exist_ok=True)
+    os.makedirs(os.path.join(args.output_dir, "latent"), exist_ok=True)
+    os.makedirs(os.path.join(args.output_dir, "prompt_embed"), exist_ok=True)
+    os.makedirs(os.path.join(args.output_dir, "prompt_attention_mask"), exist_ok=True)
+    
+    latents_json_path = os.path.join(args.output_dir, "videos2caption_temp.json")
+    with open(latents_json_path, "r") as f:
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_T5.py:30]: You've put every code inside the with, which is not necessary. You should only put the data loading part under with.

File: fastvideo/dataset/t2v_datasets.py [hunk 18]
```diff
@@ -379,6 +393,7 @@ def read_jsons(self, data):
         cap_lists = []
         with open(data, 'r') as f:
             folder_anno = [i.strip().split(',') for i in f.readlines() if len(i.strip()) > 0]
+        print(folder_anno)
         for folder, anno in folder_anno:
             with open(anno, 'r') as f:
                 sub_list = json.load(f)
```
Comment (jzhang38) [inline fastvideo/dataset/t2v_datasets.py:399]: Shoud we add sorting here as well?

File: fastvideo/utils/data_preprocess/finetune_data_T5.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+                            prompt=data['caption'],
+                        )
+                        prompt_embed_path = os.path.join(args.output_dir, "prompt_embed", video_name + ".pt")
+                        video_path = os.path.join(args.output_dir, "video", video_name + ".mp4")
+                        prompt_attention_mask_path = os.path.join(args.output_dir, "prompt_attention_mask", video_name + ".pt")
+                        # save latent
+                        torch.save(prompt_embeds[0], prompt_embed_path)
+                        torch.save(prompt_attention_mask[0], prompt_attention_mask_path)
+                        print(f"sample {video_name} saved")
+                        video = pipe.vae.decode(latent.unsqueeze(0).to(device), return_dict=False)[0]
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_T5.py:53]: We do not need to decode back anymore. It is only written there for debug.

File: fastvideo/utils/data_preprocess/finetune_data_T5.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    local_rank = int(os.getenv('RANK', 0))
+    world_size = int(os.getenv('WORLD_SIZE', 1))
+    print('world_size', world_size, 'local rank', local_rank)
+    
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    torch.cuda.set_device(local_rank)
+    if not dist.is_initialized():
+        dist.init_process_group(backend='nccl', init_method='env://', world_size=world_size, rank=local_rank)
+    
+    pipe = MochiPipeline.from_pretrained(args.model_path, torch_dtype=torch.bfloat16).to(device)
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_T5.py:21]: Change the code to only load T5, instead of using the entire pipeline.

File: fastvideo/utils/data_preprocess/finetune_data_VAE.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+        batch_size=args.train_batch_size,
+        num_workers=args.dataloader_num_workers,
+    )
+
+
+    encoder_device = torch.device(f"cuda" if torch.cuda.is_available() else "cpu")
+    torch.cuda.set_device(local_rank)
+    if not dist.is_initialized():
+        dist.init_process_group(backend='nccl', init_method='env://', world_size=world_size, rank=local_rank)
+    vae = AutoencoderKLMochi.from_pretrained(args.model_path, subfolder="vae", torch_dtype=torch.bfloat16).to("cuda")
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_VAE.py:42]: Use vae in float32, only cast the final output latents to bfloat16.

File: fastvideo/utils/data_preprocess/finetune_data_VAE.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+    os.makedirs(args.output_dir, exist_ok=True)
+    os.makedirs(os.path.join(args.output_dir, "latent"), exist_ok=True)
+    
+    json_data = []
+    for _, data in enumerate(train_dataloader):
+        video_name = os.path.basename(data['path'][0]).split(".")[0]
+        if int(video_name) % world_size != local_rank:
+            continue
+        with torch.inference_mode():
+            with torch.autocast("cuda", dtype=torch.bfloat16):
```
Comment (jzhang38) [inline fastvideo/utils/data_preprocess/finetune_data_VAE.py:53]: Use vae in float32, only cast the final output latents to bfloat16.

**Comment by jzhang38** [comment]
@rlsu9 , I will merge first. But please look at the new comments and make changes later.
